aws/data_source_aws_s3_bucket_objects.go

```diff
@@ -0,0 +1,183 @@
+
+func dataSourceAwsS3BucketObjects() *schema.Resource {
+	return &schema.Resource{
+		Read: dataSourceAwsS3BucketObjectsRead,
+
+		Schema: map[string]*schema.Schema{
+			"bucket": {
+				Type:     schema.TypeString,
+				Required: true,
+			},
+			"prefix": {
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+			"delimiter": {
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+			"encoding_type": {
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+			"max_keys": {
+				Type:     schema.TypeInt,
+				Optional: true,
+				Default:  1000,
+			},
+			"start_after": {
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+			"fetch_owner": {
+				Type:     schema.TypeBool,
+				Optional: true,
+			},
+			"arn": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"keys": {
+				Type:     schema.TypeList,
+				Computed: true,
+				Elem:     &schema.Schema{Type: schema.TypeString},
+			},
+			"common_prefixes": {
+				Type:     schema.TypeList,
+				Computed: true,
+				Elem:     &schema.Schema{Type: schema.TypeString},
+			},
+			"owners": {
+				Type:     schema.TypeList,
+				Computed: true,
+				Elem:     &schema.Schema{Type: schema.TypeString},
+			},
+		},
+	}
+}
+
+func dataSourceAwsS3BucketObjectsRead(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).s3conn
+
+	bucket := d.Get("bucket").(string)
+	prefix := d.Get("prefix").(string)
+
+	input := &s3.HeadBucketInput{
+		Bucket: aws.String(bucket),
+	}
+
+	log.Printf("[DEBUG] Reading S3 bucket: %s", input)
+	_, err := conn.HeadBucket(input)
+
+	if err != nil {
+		return fmt.Errorf("Failed listing S3 bucket object keys: %s Bucket: %q", err, bucket)
+	}
+
+	d.SetId(fmt.Sprintf("%s_%s", bucket, prefix))
+
+	arn := arn.ARN{
+		Partition: meta.(*AWSClient).partition,
+		Service:   "s3",
+		Resource:  bucket,
+	}.String()
+	d.Set("arn", arn)
+
+	listInput := s3.ListObjectsV2Input{
+		Bucket: aws.String(bucket),
+	}
+
+	if prefix != "" {
+		listInput.Prefix = aws.String(prefix)
+	}
+
+	if s, ok := d.GetOk("delimiter"); ok {
+		listInput.Delimiter = aws.String(s.(string))
+	}
+
+	if s, ok := d.GetOk("encoding_type"); ok {
+		listInput.EncodingType = aws.String(s.(string))
+	}
+
+	// "listInput.MaxKeys" refers to max keys returned in a single request
+	// (i.e., page size), not the total number of keys returned if you page
+	// through the results. "maxKeys" does refer to total keys returned.
+	maxKeys := -1
+	if max, ok := d.GetOk("max_keys"); ok {
+		maxKeys = max.(int)
+		if maxKeys > keyRequestPageSize {
+			listInput.MaxKeys = aws.Int64(int64(keyRequestPageSize))
+		} else {
+			listInput.MaxKeys = aws.Int64(int64(maxKeys))
+		}
+	}
+
+	if s, ok := d.GetOk("start_after"); ok {
+		listInput.StartAfter = aws.String(s.(string))
+	}
+
+	if b, ok := d.GetOk("fetch_owner"); ok {
+		listInput.FetchOwner = aws.Bool(b.(bool))
+	}
+
+	keys, prefixes, owners, err := listS3Objects(conn, listInput, maxKeys)
```

Comment:
It seems like we can simplify the logic by using the paginated version of the API call ([`ListObjectsV2Pages()`](https://docs.aws.amazon.com/sdk-for-go/api/service/s3/#S3.ListObjectsV2Pages)), setting `maxKeys` in the input during runtime if less than or equal to 1000, and decrementing `maxKeys` per page. e.g.

```go
var commonPrefixes []string
var keys []string
var owners []string

err := conn.ListObjectsV2Pages(listInput, func(page *s3.ListObjectsV2Output, lastPage bool) bool {
	for _, commonPrefix := range page.CommonPrefixes {
		commonPrefixes = append(commonPrefixes, aws.StringValue(commonPrefix.Prefix))
	}

	for _, object := range page.Contents {
		keys = append(keys, aws.StringValue(object.Key))

		if object.Owner != nil {
			owners = append(owners, aws.StringValue(object.Owner.ID))
		}
	}

	maxKeys = maxKeys - aws.Int64Value(page.KeyCount)

	if maxKeys <= keyRequestPageSize {
		listInput.MaxKeys = aws.Int64(int64(maxKeys))
	}

	return !lastPage
})

if err != nil {
	return fmt.Errorf("error listing S3 Bucket (%s) Objects: %s", bucket, err)
}

if err := d.Set("common_prefixes", commonPrefixes); err != nil {
	return fmt.Errorf("error setting common_prefixes: %s", err)
}

if err := d.Set("keys", keys); err != nil {
	return fmt.Errorf("error setting keys: %s", err)
}

if err := d.Set("owners", owners); err != nil {
	return fmt.Errorf("error setting owners: %s", err)
}

return nil
```

Reply:
I was using `ListObjectsV2Pages()` (in the `listS3Objects` func) but not with the anonymous function. I didn't realize you could do that. This is cool.

Reply:
🎉 Yeah the paginated functions are great.

Another way this could be even further simplified is also creating slices of `[]*s3.CommonPrefix` and `[]*s3.Object` which get handled in the anonymous pagination function then grab out the information of what we need afterwards, but that does add some (slight) compute/memory overhead to get the desired end results:

```go
var commonPrefixes []*s3.CommonPrefix
var objects []*s3.Object

err := conn.ListObjectsV2Pages(listInput, func(page *s3.ListObjectsV2Output, lastPage bool) bool {
	commonPrefixes = append(commonPrefixes, page.CommonPrefixes...)
	objects = append(objects, page.Contents...)

	maxKeys = maxKeys - aws.Int64Value(page.KeyCount)

	if maxKeys <= keyRequestPageSize {
		listInput.MaxKeys = aws.Int64(int64(maxKeys))
	}

	return !lastPage
})
// ...
```

Either which way is totally fine I think. Just noting this as well since it is a pattern we do use in some other places in the Terraform AWS Provider in case you run across it.